Simplify local reproduction instructions

angerson opened the pull request:
This change is a big simplification of reproducing a build locally ([see it here](https://github.com/angerson/tensorflow/tree/master/ci/official)):

- Moved a lot of configurable behavior into source-able local_* env files
- Deleted a lot of extraneous instructions
- Added "how do I choose..." instructions for clarification
- Fixed a bug preventing multiple invocations of rename_and_verify_wheels
- Made any.sh and bisect.sh use the cleaner method of sourcing local_ envs
- Cleaned up ci_default and added "how to see an overview of env variables"

One of the big things is that I added some logic in local_default that resets the list of Bazel common args to remove things like --config=rbe. This way a user can choose to run a "release" env configuration without being blocked by permission errors, or by needing to manually amend their configuration. As a result, gathering the correct set of config values is quite easy (see the new instructions). The alternative would be to split "auth-related" flags into a separate variable, which I don't like, as I think it's more confusing for future maintainers to have to continuously decide which options go where.

angerson committed:
Simplify local reproduction instructions

This change is a big simplification of reproducing a build locally:

- Moved a lot of configurable behavior into source-able local_* env files
- Deleted a lot of extraneous instructions
- Added "how do I choose..." instructions for clarification
- Fixed a bug preventing multiple invocations of rename_and_verify_wheels
- Made any.sh and bisect.sh use the cleaner method of sourcing local_ envs
- Cleaned up ci_default and added "how to see an overview of env variables"

One of the big things is that I added some logic in local_default that resets
the list of Bazel common args to remove things like --config=rbe. This way a
user can choose to run a "release" env configuration without being blocked by
permission errors, or by needing to manually amend their configuration. As a
result, gathering the correct set of config values is quite easy (see the new
instructions). The alternative would be to split "auth-related" flags into
a separate variable, which I don't like, as I think it's more confusing for
future maintainers to have to continuously decide which options go where.

MichaelHudgins reviewed (commented):
This overall looks good (and way easier for external re-use). 

Which combination of local options did you end running yourself?

angerson committed:
Switch to more approachable example starter

angerson committed:
Fix RBE settings and add root warnings

angerson committed:
Docker is enabled by default on linux

angerson committed:
Rename docker.sh and fix any.sh and local vars

angerson committed:
Tweak any.sh

angerson committed:
Note about Docker

angerson committed:
More docker instructions

angerson commented:
Ok, this is ready for review again. I fixed a few issues, including one bug with including TFCI variables from the environment, and added more documentation -- mostly about docker. I have verified that (on my Linux machine), these all appeared to be working:

- local_multicache
- local_rbe
- local_nodocker
- pycpp.sh
- any.sh
- wheel.sh

I don't have a good way to verify bisect.sh yet but I think it's fixed. I'll check on it again after this gets merged.

MichaelHudgins commented on ci/official/README.md:
```diff
@@ -1,295 +1,131 @@
+    - Uses `pycpp.sh`, `code_check_changed_files.sh`
+    - Also uses `envs/continuous_...`
+
+These `envs` match up with an environment matrix that roughly covers:
+
+-   Linux, MacOS, and Windows machines (these pool definitions are internal)
+-   x86 and arm64
+-   CPU-only, or with NVIDIA CUDA support (Linux only), or with TPUs
+-   Different Python versions
```
WDYT about making this just an explicit table?  This table on JAX is what made me think about it https://github.com/google/jax?tab=readme-ov-file#supported-platforms

angerson replied:
Sure, but let's punt the table for later, since some of the jobs right now are WIP.

MichaelHudgins commented on ci/official/README.md:
```diff
@@ -1,295 +1,131 @@
+## TensorFlow's Official CI and Build/Test Scripts
+
+TensorFlow's official CI jobs run the scripts in this folder. Our internal CI
+system, Kokoro, schedules our CI jobs by combining a build script with an `env` file that is filled with configuration options:
```
Hmm, maybe worth stating that the env files correspond to the envs folder in this directory.  I know most people will notice but i could see a few asking "What env files"?

angerson replied:
Good idea, tweaked the wording here.

MichaelHudgins reviewed (commented):
This is looking really good.  I have a few things to get through but i will try and run these instructions on my mac to see if they work for a macos flow later today.

angerson committed:
Tweak readme

angerson committed:
Fix code_check_changed_files

angerson committed:
Remove go link

angerson committed:
Remove suggested target example
